test/powershell/Language/Scripting/PipelineBehaviour.Tests.ps1

```diff
@@ -0,0 +1,122 @@
+Describe 'Function Pipeline Behaviour' -Tag 'CI' {
```

Comment:
I know we currently don't have any testing like this (to my knowledge) and the onus shouldn't be on you to provide it, but it would be really nice to test a binary cmdlet with a `Dispose()` implementation. Particularly some badly implemented ones that do things like throw or block or something

Reply:
I will look to throw something together. Honestly I'm not really clear on what the defined behaviour expected for cmdlets throwing from Dispose() actually is. There _is_ code and a few scant comments mentioning it in CommandProcessorBase but it's clear on what's expected to happen, other than that throwing is an expected possibility.

Reply:
> Honestly I'm not really clear on what the defined behaviour expected for cmdlets throwing from Dispose() actually is

Yeah it's a very good question. My suspicion is that nobody is very clear and that we are now at that point where we need to be, since we're exposing it much more as something to think about.

Reply:
Coming back to this after some thought.

If we consider this kind of C# pattern:

```cs
try
{
    throw new NotSupportedException();
}
finally
{
    throw new NotImplementedException();
}

```

From what I can tell, the exception in the `finally{}` completely overrides the one thrown from the `try`; I can't see a way to even look at the one that comes from the `try` after a bit of experimentation in sharplab.

But something up-scope can still catch it, provided they're able to catch the NotImplementedException(). Doesn't look like you can still catch the NotSupportedException() after that point from what I could see.

If we were to accept that as "expected" more or less, since `Dispose()` / `cleanup{}` is executed in a finally, if it throws we would be OK just to report the last error.

Current code here does one extra bit better in that both errors will show up in `$error` so if you do really need that initial error to help diagnose things, you still get both... but only the terminating error from the dispose is reported directly.

Reply:
> From what I can tell, the exception in the finally{} completely overrides the one thrown from the try; I can't see a way to even look at the one that comes from the try after a bit of experimentation in sharplab.

Yeah I think that's the case. Same as with something like:

```csharp
try
{
    return 1;
}
finally
{
    return 2;
}
```

The runtime sets a value to be returned/thrown in the try block and then overwrites that value in the finally block.

But my question is more about the calling context of the dispose block. Does the pipeline terminate? Does the script terminate?

My assumptions would be:

- We call dispose/cleanup exactly once for every command in every pipeline
- It's always called last, when the pipeline is cleaned up -- the only thing that could be called after the dispose is another cmdlet's dispose method
- We don't guarantee the state of the cmdlet or the pipeline when it's called, other than that the cmdlet is fully constructed
- The question remains in my mind as to which properties/APIs on the cmdlet/function are valid when it's being disposed:
    - I would assume the PSCmdlet APIs all work
    - I assume `WriteObject()` does nothing
    - I assume `WriteInformation()`, `WriteWarning()`, `WriteError()` all just write to the streams as normal
    - I'm guessing that `ThrowTerminatingError()` will elevate the pipeline cleanup to an error, possibly replacing an existing error

But I'm not sure how many of those things make sense, are possible, or will have corner cases.

Reply:
As far as I can tell, those assumptions are completely correct. Because we prevent APIs from detecting the pipeline is stopping, they're all perfectly valid and can all be called safely.

Writing to streams with commands or by "dropping" output back to the pipeline (which results in an internal WriteObject() call if I'm not mistaken?) work just fine, I'll have to check if it somehow breaks when you directly call the $PSCmdlet.Write*() APIs but I believe it should work the same.

ThrowTerminatingError() _does_ replace the existing error. The only point pursuant to that which is a little different than might be expected is that **both** the original error and the new error are written to `$error`. However, when catching such a terminating error arising from that kind of state, you'll only be able to `catch {}` the error from the `cleanup{}` block. Personally I think that's good, because that way you have proper insight into what actually happened there, and it lets both the function and the caller handle the error state as they need to.